src/object.d

```diff
@@ -3190,11 +3199,24 @@ void __ctfeWriteln(T...)(auto ref T values) { __ctfeWrite(values, "\n"); }
  * Create RTInfo for type T
  */
 
+template RTInfoImpl(size_t[] pointers)
+{
+    immutable size_t[pointers.length] data = pointers[];
+    immutable RTInfoImpl = data.ptr;
```

Comment:
> I'd suggest to convert the size_t[] first to a ubyte[] array (while leaving out all 0 values at the end), then use Variable-length quantity (Base128) to store the array length in front of the metadata, and store it as a ubyte*.

That's a possible optimization, but i'm not sure the extra decoding is worth it. There are only a limited number of types in the binary, and the few extra kB was not really a problem so far (I'm only using this with Visual D, though).

> ... and removes one indirection.

RTInfo is already a pointer into the buffer, and the compiler/linker is free to omit the (length,ptr) pair. Or do you have another indirection in mind?

> For any struct below 64-byte this would only require 8+1+1=10 bytes instead of 8+16+8=32 bytes. 

16 bytes. The 10 bytes are likely aligned to 16 bytes anyway.

> Truncating zeros at the end will save on structs with big buffers as last field.

This could be a nice optimization, but we loose the easy accessible information of the size of a struct that is repeated in an array.

> We should also use null pointers for data that doesn't require scanning, 

Allocations that don't require scanning have the NO_SCAN flag set, so the info isn't touched anyway. The current implementation already has the special pointers `rtinfoNoPointers` and `rtinfoHasPointers`, but more as compiler bug workarounds.

> and could use 0 length as an escape magic preceeding more complex encodings, e.g. run length encoding for structs with big nogc holes.

This could help with the following:

I think the more important part is to optimize the way the bitmap pointers are stored alongside the allocated memory:

- for arrays only one copy of the RTInfo data needs to be copied
- for large allocations we might just store the pointer to the RTInfo
- if we know the size of array allocations, we could skip scanning the unused space (up to 50% of the allocation)

This requires more information being passed to the GC (e.g. what is an array, array sizes, etc.). It also doesn't fit with the current `gc_emplace` functionality. Maybe we can restrict the latter to change the type description of a full allocation.